src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1868,7 +1901,16 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
 
     if (block.deleteMarker == false) { // provided version is not a delete marker and contains data
       if (source->get_bucket()->versioned()) { 
-	if (blockDir->del(dpp, &block, y) == 0) { // delete versioned head object
+        if (block.cacheObj.objName != source->get_name()) { // else the head object key will match the versioned head object key in the directory 
+	  if (!nullDelete) {
+	    ret = blockDir->del(dpp, &block, y); // delete versioned head object
+	    if (ret < 0 && ret != -ENOENT) {
+	      ldpp_dout(dpp, 0) << "Failed to delete versioned head object for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
+	      return ret;
+	    }
+	  } // else, only delete the data blocks since the null head object directory entry was updated to a delete marker
+	  
+	  ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): deleting versioned head object data block for " << head_oid_in_cache << dendl;
```

Comment:
again, below cache blocks are being deleted without a check whether they are dirty or not. We had decided that non-dirty/ clean blocks will be taken care of during eviction